clang/docs/checkedc/rfc/Include-Checked-Hdrs.md

```diff
@@ -0,0 +1,187 @@
+       the flag `-DIMPLICIT_INCLUDE_CHECKED_HDRS` is passed to the compilation
+       of all source files through the `CFLAGS` variable.
+
+     - **Cons**: 1) In some cases integration with the build system will be more
+       involved if the above flag needs to be passed to the compilation of most
+       source files and avoided for a few source files that may want to directly
+       include system header files in order to avoid Checked-C-specific
+       declarations. 2) It is difficult to accomodate clang-specific
+       declarations.
+
+## Proposed Solution:
+
+The proposed solution is as follows:
+  - It is a hybrid solution based on ideas in Solutions 3 and 1 that
+    handles the presence of both clang-specific and Checked-C-specific
+    declarations in header files. It also incorporates the feedback from CCI to
+    make implicit inclusion of checked header files as default.
+  - We primarily adopt Solution 3 for header files that do not have
+    clang-specific declarations, but have Checked-C-specific declarations (there
+    are 13 such header files at present). To implicitly include checked headers
+    by default, the compilation flag (as proposed in Solution 3) now gives a way
+    to **opt out** of implicit inclusion. Accordingly, it is renamed to
+    NO_IMPLICIT_INCLUDE_CHECKED_HDRS.
+      - For each system header file, say `foo.h`, that does not have
+        clang-specific declarations but has Checked-C-specific declarations, we
+        will add a new file also called `foo.h` that will contain the following:
+
+            #ifdef NO_IMPLICIT_INCLUDE_CHECKED_HDRS
+            #include_next <foo.h>
+            #else
+            #include <foo_checked.h>
+            #endif
+
+      - The file `foo_checked.h` will contain `#include_next <foo.h>` and the
+        Checked-C-specific declarations.
+
+  - For a system header file that contains clang-specific declarations and also
+    Checked-C-specific declarations (there is one such header file at present),
+    we will do the following:
+      - Let `bar.h` be such a header file. Then there already 
+        exists a file called `bar.h` in the `clang/lib/Headers` directory,
+        which contains clang-specific declarations in the following format:
+
+            #include_next <bar.h>
+            <currently existing clang-specific declarations>
+
+        At the end of this pre-existing `bar.h`, we will add the following:
+
+            #ifndef NO_IMPLICIT_INCLUDE_CHECKED_HDRS
+            #include <bar_checked_internal.h>
+            #endif
+
+      - All the Checked-C-specific declarations (that are currently present in
+        `bar_checked.h`) will be moved to `bar_checked_internal.h`. The file
+        `bar_checked.h` will be modified to contain just the following:
+
+            // Force the inclusion of Checked-C-specific declarations
+            #undef NO_IMPLICIT_INCLUDE_CHECKED_HDRS
+            #include <bar.h>   
```

Comment:
If you do keep this approach rather than taking my suggestion above, might the following be safer?
```suggestion
            #include <bar.h>
            #include <bar_checked_internal.h>
```
`bar_checked_internal.h` might get scanned twice, so it would need a header guard and there might be a slight performance cost, but it would make it possible for a user to set `NO_IMPLICIT_INCLUDE_CHECKED_HDRS` for one `#include <foo.h>` directive in a translation unit without the risk of an `#include <bar_checked.h>` earlier in the translation unit clobbering `NO_IMPLICIT_INCLUDE_CHECKED_HDRS` (if that is a scenario you care about).

Reply:
@mattmccutchen-cci Thank you for your suggestion - you are right! I have incorporated it in the latest commit.